### ISSUE TITLE
[Bitmex] open orders amount is relative to the quote instead of the base (maybe any order type)

### ISSUE THREAD

**Issue by AoDev** [ccxt/ccxt#4081]
- OS: **OSX**
- Programming Language version: **javascript**
- CCXT version: **1.17.439**
- Exchange: **Bitmex**
- Method: **fetchOpenOrders**

Comparing Bitmex with let's say Binance or Kraken, orders amount are usually corresponding to the `base` of the market symbol.

Eg: Binance open order

```
{
  "symbol": "ZIL/BTC",
  "type": "limit",
  "side": "sell",
  "price": 0.00000668,
  "amount": 4545
}
```
Which translates to "selling 4545 ZIL at a price of 0.00000668 BTC".

Now, with Bitmex:

```
{
  "symbol": "BTC/USD",
  "type": "stoplimit",
  "side": "sell",
  "price": 6050,
  "amount": 4400
}
```
Which, if we translate again: "selling 4400 BTC at a price of 6050 USD"

But on Bitmex, the real meaning is that we reduce (or increase) our position of 4400 USD if the price touches 6050 USD.

I checked the Bitmex API response and I know it's really not ideal. At least in the `info` part, there is no details of margin, leverage etc. 

But the current mapping seems wrong, or is it not? Your thoughts? 
In my own app, when I saw "Buy a 1000 BTC order, I had a shock until I understood" :D

**Comment by kroitor** [comment]
The mapping may be wrong, indeed. From their docs it's not very clear, basically... The order parser is right here: https://github.com/ccxt/ccxt/blob/master/js/bitmex.js#L437 

Yep, it most likely has mistakes – with `cumQty`←→`filled/cost`, the amount and probably some others fields. Feel free to add edits to it for a better mapping, we appreciate all help, your PRs are welcome )

Also, this may be related: https://github.com/ccxt/ccxt/pull/3779

**Comment by AoDev** [comment]
Ok, I will first check their API response and make a "report" about what we have and what we don't have.

**Comment by AoDev** [comment]
So thinking out loud, please share your thoughts on this. 

From an isolated order point of view, we could simply divide the amount of contracts ["BCT/USD"] by the price ["USD"] to get the actual amount of base ["BTC"].

Example with `BTC/USD`
Open order from Bitmex API:

```
side: "Sell"
orderQty: 4400 // <- this is the amount of contracts in USD,
price: 6050
```
This translates roughly to "Sell `4400/6050` BTC @ price of 6050 USD

So, could be mapped to:

```
side: "Sell"
amount: 0.72 // <- these are BTC or the "base"
price: 6050
```

Note that orders like these don't have any information about the context, like the current position and leverage.

So, while this fix would work, but give me your opinion @kroitor, people must be aware that they must interpret the data in the context of their position, leverage, etc.

**Comment by kroitor** [comment]
@AoDev one likely issue with the above is that it may only be true for BTC/USD and not the rest of their contracts, I'm afraid. Have you tried any other pair? That is explained in the linked post above, we might have to bring `tickSize`, `lotSize` and `multiplier` into this calculation.

Also, because of how order matching works with precision and rounding, we can't really divide the cost by price and assume that we have an actual correct amount from there, in other words, actual amounts are rounded not truncated, so the true value might as well be 0.73 in the above example case. 

Because division yields "side effects" of that sort, we'd rather set the `amount` to `undefined/None/null` to indicate an ambiguity to the user and would set just the `price` to 6050 and just the cost to `4400`. We think, from there, the user would then decide what to do with a missing amount and whether he wants to derive it from other fields or not. 

In the context of your initial posts, that will ultimately mean less confusion, compared to making up fields that aren't provided by the exchange or don't fit the order structure due to specifics of margin/position/leveraged trading. And, in the end, we might as well introduce a structure like `position` alongside the `order` to get our hands on it and to model for those specifics properly.

Our first and foremost priority at this point is making sure that the above approach will work seamlessly across all their markets. If you can proceed on that, that will be fantastic! ;)

**Comment by AoDev** [comment]
I didn't check with the other contracts, you are right, I'll see.

> in the end, we might as well introduce a structure like position alongside the order to get our hands on it and to model for those specifics properly

That's something I had in mind, and wanted to ask, more a thing for the future. We definitely need a structure for margin trading. (Note in Bitmex docs, they refer to [some standard ](https://www.onixs.biz/fix-dictionary/5.0.SP2/msgType_D_68.html).

On the other hand, maybe a concept of "extrapolated" could be introduced (not sure about the name).. a key in the data where, when facing every exchange intricacies, we could add "extrapolated data". People could use it at their own perils so to speak.

Could result in this:

```
{
   amount: NaN, // people expect a number so maybe this is more explicit
   side: 'sell',
   price: 6050,
   cost: 4400, // <- right now I see it's set to 0
   extrapolated: {
      amount: 0.72727272 // <- Careful if you use this
   }
}
```
This can be extended to any field so very flexible for the future. I don't know if there are many cases where this would be useful. But it's a separate idea, I don't want to complicate things.

I checked the other post and it's basically the same issue. "trade volume" = "amount"

**Comment by kroitor** [comment]
Closing this for now if you don't mind, feel free to reopen it if needed.
